### PR TITLE
scala-codegen: add support for enum type

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR continue the effort to add `Enum` type to the SDK (See issue #105)

This PR add `Enum` type support for the scala and codegen .  

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala [hunk 1]
```diff
@@ -244,20 +242,26 @@ object CodeGen {
...
+
+    val records = definitions.collect {
+      case sdt @ ScopedDataType(qualName, typeVars, r: Record[RT]) => sdt copy (dataType = r)
+    }
+
+    val variants = definitions.collect {
+      case sdt @ ScopedDataType(qualName, typeVars, v: Variant[VT]) => sdt copy (dataType = v)
+    }
+
+    val enums = definitions.collect {
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]: I'd rather use `partitionEithers` twice to preserve the exhaustiveness checking, and continue returning the specific types from `splitNTDs`.

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala [hunk 2]
```diff
@@ -268,16 +272,24 @@ object CodeGen {
...
+      records.map {
+        case ddt @ ScopedDataType(name, vars, _) => (name -> vars.toList) -> ddt
+      }(breakOut)
 
     val noDeletion = Set.empty[(Identifier, List[Ref.Name])]
     // both traverseU can change to traverse with -Ypartial-unification
     // or Scala 2.13
     val (deletedRecords, newVariants) =
-      variants.traverseU {
+      variants.toList.traverseU {
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]:
```suggestion
      variants.traverseU {
```

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/DamlDataTypeGen.scala [hunk 3]
```diff
@@ -134,11 +135,56 @@ object DamlRecordOrVariantTypeGen {
...
+            q"""
+            object ${TermName(className)} extends
+              ${tq"$domainApiAlias.EnumCompanion[$appliedValueType]"} {
+              ..${constructors.zipWithIndex.map {
+              case (c, i) =>
+                q"""case object ${TermName(c.capitalize)} extends $appliedValueType($c, $i) """
+            }}
+            
+              val firstValue: $appliedValueType = ${TermName(firstValue.capitalize)}
+              val otherValues:  Vector[$appliedValueType] =
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/DamlDataTypeGen.scala]: Add `stdVectorType` alongside `stdMapType` to `LFUtil`, and

```suggestion
              val otherValues:  $stdVectorType[$appliedValueType] =
```

Comment (remyhaemmerle-da) [reply language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/DamlDataTypeGen.scala]: fixed

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]: fixed

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala [hunk 4]
```diff
@@ -43,6 +45,13 @@ abstract class ShrinkEncoding extends LfTypeEncoding {
 
   override def variant[A](variantId: rpcvalue.Identifier, cases: VariantCases[A]): Out[A] = cases
 
+  override def enum[A](enumId: rpcvalue.Identifier, cases: EnumCases[A]): Out[A] = cases
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala]: not applied anymore

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala [hunk 5]
```diff
@@ -32,6 +32,8 @@ abstract class ShrinkEncoding extends LfTypeEncoding {
 
   type VariantCases[A] = Shrink[A] // probably
 
+  type EnumCases[A] = Shrink[A]
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala]: does not apply anymore

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShowEncoding.scala [hunk 6]
```diff
@@ -80,6 +85,11 @@ object ShowEncoding extends ShowEncoding {
     }
   }
 
+  class EnumCasesImpl extends Plus[EnumCases] {
+    def plus[A](a: Show[A], b: => Show[A]): Show[A] = new Show[A] { (f: A) => a.show(f) ++ b.show(f)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShowEncoding.scala]: does not apply anymore

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShowEncoding.scala [hunk 7]
```diff
@@ -46,23 +48,26 @@ abstract class ShowEncoding extends LfTypeEncoding {
...
   override def variantCase[B, A](caseName: String, o: Out[B])(inject: B => A)(
-      select: A PartialFunction B): VariantCases[A] = new Show[A] {
-
-    override def show(a: A): Cord = {
-      select.lift(a).map(b => o.show(b)) match {
-        case Some(b) => Cord(caseName, "(", b, ")")
-        case None => Cord.empty
-      }
+      select: A PartialFunction B): VariantCases[A] =
+    new Show[A] { (a: A) =>
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShowEncoding.scala]: does not apply anymore

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShowEncoding.scala [hunk 8]
```diff
@@ -46,23 +48,26 @@ abstract class ShowEncoding extends LfTypeEncoding {
 
   override def fields[A](fi: Field[A]): RecordFields[A] = fi
 
+  override def enum[A](enumId: Identifier, cases: EnumCases[A]): Out[A] = cases
+
+  override def enumCase[A](caseName: String)(a: A): EnumCases[A] = new Show[A] { (b: A) =>
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShowEncoding.scala]: does not apply anymore

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/EqualityEncoding.scala [hunk 9]
```diff
@@ -28,6 +30,11 @@ abstract class EqualityEncoding extends LfTypeEncoding {
 
   override def fields[A](fi: Field[A]): RecordFields[A] = fi
 
+  def enum[A](enumId: Identifier, cases: EnumCases[A]): Out[A] = cases
+
+  override def enumCase[A](caseName: String)(a: A): EnumCases[A] =
+    (a1: A, a2: A) => a == a1 && a == a2
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/EqualityEncoding.scala]: fixed

File: language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/encoding/LfTypeEncoding.scala [hunk 10]
```diff
@@ -123,6 +146,12 @@ object LfTypeEncoding {
     override def fields[A](fi: Field[A]): RecordFields[A] =
       (fst.fields(fi._1), snd.fields(fi._2))
 
+    override def enum[A](enumId: rpcvalue.Identifier, cases: EnumCases[A]): Out[A] =
+      (fst.enum(enumId, cases._1), snd.enum(enumId, cases._2))
+
+    override def enumCase[A](caseName: String)(a: A): EnumCases[A] =
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/encoding/LfTypeEncoding.scala]: does not apply anymore

File: language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    override def write(enum: T): rpcvalue.Value.Sum =
+      rpcValues(isEnum(enum).index)
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  implicit final val `the enum LfEncodable`: LfEncodable[T] = new LfEncodable[T] {
+    override def encoding(lte: LfTypeEncoding): lte.Out[T] =
+      lte.enumAll(
+        ` dataTypeId`,
+        (firstValue +: otherValues).map(isEnum(_).constructor).zipWithIndex.toMap,
```
Comment (S11001001) [inline language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala]:
```suggestion
        a => a.index,
```

File: language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        OneAnd(firstValue, otherValues).map(x => isEnum(x).constructor -> x))
```
Comment (S11001001) [inline language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala]: Doesn't refer to `lte`, so can be pulled out like `readers` and `rpcValues` (with the same `lazy val` for the containing `val`).

File: language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala [hunk 13]
```diff
@@ -405,6 +400,24 @@ object LfTypeEncodingSpec {
...
     }
 
+    override def enumAll[A](
+        enumId: rpcvalue.Identifier,
+        index: A => Int,
+        cases: OneAnd[Vector, (String, A)],
+    ): Out[A] =
+      new Value.InternalImpl[A] {
+        private val readerMap =
+          cases.toVector.toMap
```
Comment (S11001001) [inline language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala]:
```suggestion
          cases.toVector
```

File: language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala [hunk 14]
```diff
@@ -405,6 +400,24 @@ object LfTypeEncodingSpec {
...
+
+        override def read(av: VSum): Option[A] =
+          av.enum.flatMap(e => readerMap.get(e.constructor))
```
Comment (S11001001) [inline language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala]:
```suggestion
          av.enum.flatMap(e => readerMap.lift(e.index))
```

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/DamlDataTypeGen.scala [hunk 15]
```diff
@@ -134,11 +135,56 @@ object DamlRecordOrVariantTypeGen {
...
+      val klass =
+        q"""
+          sealed abstract class ${TypeName(className)}(
+            override val constructor: String,
+            override val index: Int
+          ) extends ${tq"$domainApiAlias.EnumRef"} {
+            ..$rootClassChildren
+          }"""
+
+      val (imports, companionObject) = constructors match {
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/DamlDataTypeGen.scala]: For whatever reason, the `::` pattern is super brittle.

```suggestion
      val (imports, companionObject) = (constructors: List[Ref.Name]) match {
```

File: language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala (hunk 11 above)
Comment (remyhaemmerle-da) [reply language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala]: Indeed. fixed.

File: language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala (hunk 14 above)
Comment (S11001001) [reply language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala]: Oops n/m

File: language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala (hunk 13 above)
Comment (S11001001) [reply language-support/scala/codegen-testing/src/test/scala/com/digitalasset/ledger/client/binding/encoding/LfTypeEncodingSpec.scala]: Oops n/m

File: language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala (hunk 12 above)
Comment (remyhaemmerle-da) [reply language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala]: Done.

File: language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala [hunk 16]
```diff
@@ -43,6 +44,15 @@ abstract class ShrinkEncoding extends LfTypeEncoding {
 
   override def variant[A](variantId: rpcvalue.Identifier, cases: VariantCases[A]): Out[A] = cases
 
+  override def enumAll[A](
+      enumId: Identifier,
+      index: A => Int,
+      cases: OneAnd[Vector, (String, A)],
+  ): Out[A] =
+    Shrink { a: A =>
+      Stream(a).filter(index(_) == 0)
```
Comment (S11001001) [inline language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala]:
```suggestion
      if (index(a) == 0) Stream.empty else Stream(cases.head._2)
```

Comment (remyhaemmerle-da) [reply language-support/scala/codegen-testing/src/main/scala/com/digitalasset/ledger/client/binding/encoding/ShrinkEncoding.scala]: fixed

**Pull request merged**